### DISCUSSION TITLE
How to deal with packages not in Alma repos (manylinux_2_28)

### DISCUSSION THREAD

**Discussion by drebbe-intrepid** [pypa/manylinux#1617] [Q&A]
I'm trying to figure out the best route to take here. I have a new python library I'm trying to get automated building working with manylinux. My issue comes from library dependencies not available in Alma Linux. 

Here is a list of dependencies (on Debian/Ubuntu): `libsfml-dev libudev-dev libftdi1-dev`

I'm using PyO3 and maturin as the build system.

I believe I have two ways I can go about this:

1. Manually build and compile all packages through the docker image (not sure how to do this, custom script to do it all?)
2. Get a custom image of manylinux based on a different distro like Debian or Ubuntu (not sure how to do this either).

edit: Architectures are aarch64 (arm64), x86, and x86_64.

Thoughts?

**Comment by mayeut** [comment]
1. If you can rebuild packages from sources, then you'd more or less become distro agnostic (depending on the internals of those packages). libudev-dev has been replaced by systemd-devel on RHEL derivatives (such as AlmaLinux). This seems a bit close to the distro though.

2. You can build a manylinux like image based on an ubuntu base image and installing multiple versions of python using the [deadsnakes ppa](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa), you will need to [add some tools](https://github.com/pypa/manylinux/blob/7c7239e8db5047145c60c40d2474b7f5c69781be/docker/build_scripts/finalize.sh#L45-L86) (like auditwheel). Depending on which tools you're using, you might need to create some links for those tools to find python where they [expect it to be in a manylinux image](https://github.com/pypa/manylinux/tree/main?tab=readme-ov-file#image-content).

**Reply by drebbe-intrepid** [reply]
Thanks for the reply. I've still been struggling with this due to needing aarch64 architecture also. I'm at the point now where I'm just going to setup a self-hosted arm runner and have a docker image (more than likely based on debian 12). I'm going to evaluate pyenv for the different python versions, I think if needed.

I'm using rust and maturin so I think I can just target an older version since it will build with ABI3 compat.
